lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -48,6 +48,182 @@ void _setStaticStyleAttributes(html.HtmlElement domElement) {
+  elementStyle
+    ..whiteSpace = 'pre-wrap'
+    ..alignContent = 'center'
+    ..padding = '0'
+    ..opacity = '1'
+    ..color = 'transparent'
+    ..backgroundColor = 'transparent'
+    ..background = 'transparent'
+    ..outline = 'none'
+    ..border = 'none'
+    ..resize = 'none'
+    ..textShadow = 'transparent'
+    ..transformOrigin = '0 0 0';
+
+  /// This property makes the input's blinking cursor transparent.
+  elementStyle.setProperty('caret-color', 'transparent');
+}
+
+/// Form that contains all the fields in the same AutofillGroup.
+///
+/// These values are to be used when autofill is enabled and there is a group of
+/// text fields with more than one text field.
+class AutofillGroup {
+  AutofillGroup(
+      {this.formElement, this.elements, this.items, this.singleElement});
+
+  final html.FormElement formElement;
+
+  final Map<String, html.HtmlElement> elements;
+
+  final Map<String, Autofill> items;
+
+  final bool singleElement;
+
+  factory AutofillGroup.fromFrameworkMessage(
+    Map<String, dynamic> focusedElementAutofill,
+    List<dynamic> fields,
+  ) {
+    final singleElement = (fields == null);
+    final Autofill focusedElement =
+        Autofill.fromFrameworkMessage(focusedElementAutofill);
+    final Map<String, html.HtmlElement> elements =
+        Map<String, html.HtmlElement>();
+    final Map<String, Autofill> items = Map<String, Autofill>();
+    final html.FormElement formElement = html.FormElement();
+
+    // Don't validate the fields on the form.
+    formElement.noValidate = true;
+
+    _hideAutofillElements(formElement);
+
+    if (!singleElement) {
+      for (Map<String, dynamic> field in fields) {
+        final Map<String, dynamic> autofillInfo = field['autofill'];
+        final Autofill autofill = Autofill.fromFrameworkMessage(autofillInfo);
+
+        // The main text editing element will not be created here.
+        if (autofill.uniqueIdentifier != focusedElement.uniqueIdentifier) {
+          EngineInputType engineInputType =
+              EngineInputType.fromName(field['inputType']['name']);
+
+          html.HtmlElement htmlElement = engineInputType.createDomElement();
+          autofill.editingState.applyToDomElement(htmlElement);
+          autofill.applyToDomElement(htmlElement);
+          _hideAutofillElements(htmlElement);
+
+          items[autofill.uniqueIdentifier] = autofill;
+          elements[autofill.uniqueIdentifier] = htmlElement;
+          formElement.append(htmlElement);
+        }
+      }
+    }
+
+    return AutofillGroup(
+      formElement: formElement,
+      elements: elements,
+      items: items,
+      singleElement: singleElement,
+    );
+  }
+
+  void placeForm(html.HtmlElement mainTextEditingElement) {
+    formElement.append(mainTextEditingElement);
+    domRenderer.glassPaneElement.append(formElement);
+  }
+
+  void removeForm() {
+    formElement.remove();
+  }
+
+  List<StreamSubscription<html.Event>> addEventListeners() {
```

Comment:
This method seems to have a non-trivial lifecycle. For example, calling it twice will result in duplicate event subscriptions. Calling it after `removeForm` will subscribe to detached elements.

I would try to express this functionality with a simpler lifecycle. For example, it seems like the list of elements does not change after the constructor call. Which means that if the list of form fields changes we tear down this class and create a new instance. Perhaps then we could initialize the subscriptions in the constructor once and cancel subscriptions in `removeForm`?

If a simple lifecycle is not possible, then I would recommend documenting it. As an example, here's another class with a complex lifecycle: https://github.com/flutter/engine/blob/45e6cfd1ffcdf3f380bbf3f9bdab61ae68e93302/lib/web_ui/lib/src/engine/text/ruler.dart#L303.

Reply:
Done. Thanks for the suggestion. I put a lot of thought behind it to address your concern. On the other hand detaching the lifecycle management for this method from rest of all the listener subscriptions seemed more complicated in the end.

I added a more detailed comment. PHAL.